esphome/components/esp32_improv/esp32_improv_component.h

```diff
@@ -42,6 +46,12 @@ class ESP32ImprovComponent : public Component, public BLEServiceComponent {
   void stop() override;
   bool is_active() const { return this->state_ != improv::STATE_STOPPED; }
 
+#ifdef USE_ESP32_IMPROV_STATE_CALLBACK
+  void add_on_state_callback(std::function<void(improv::State)> &&callback) {
+    this->state_callback_.add(std::move(callback));
+  }
+#endif
```

Comment:
```suggestion
  void add_on_state_callback(std::function<void(improv::State, improv::Error)> &&callback) {
    this->state_callback_.add(std::move(callback));
  }
```
Optional Allow please sending the error state to the events as well.